#https://www.kaggle.com/esercicek/iris-dataset-compare-classification-models

In [2]:
#import libraay

import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
import numpy as np # linear algebra
import matplotlib.pyplot as plt
import seaborn as sns


In [3]:
df = pd.read_csv( r"C:\Users\DELL\Downloads\Iris.csv")
df.head()

Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm      Species
0   1            5.1           3.5            1.4           0.2  Iris-setosa
1   2            4.9           3.0            1.4           0.2  Iris-setosa
2   3            4.7           3.2            1.3           0.2  Iris-setosa
3   4            4.6           3.1            1.5           0.2  Iris-setosa
4   5            5.0           3.6            1.4           0.2  Iris-setosa

In [4]:
#species weill be in numeric(changing category variable to  numeric)
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
df.Species = le.fit_transform(df.Species)
df.head()

Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm  Species
0   1            5.1           3.5            1.4           0.2        0
1   2            4.9           3.0            1.4           0.2        0
2   3            4.7           3.2            1.3           0.2        0
3   4            4.6           3.1            1.5           0.2        0
4   5            5.0           3.6            1.4           0.2        0

In [5]:
df.shape  #data se have 150 rows and 6 columns


(150, 6)

In [6]:
df.Species.value_counts()

2    50
1    50
0    50
Name: Species, dtype: int64

In [7]:
#check the datatypes
df.dtypes

Id                 int64
SepalLengthCm    float64
SepalWidthCm     float64
PetalLengthCm    float64
PetalWidthCm     float64
Species            int32
dtype: object

In [8]:
#check null value
df.isnull().sum()

Id               0
SepalLengthCm    0
SepalWidthCm     0
PetalLengthCm    0
PetalWidthCm     0
Species          0
dtype: int64

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             150 non-null    int64  
 1   SepalLengthCm  150 non-null    float64
 2   SepalWidthCm   150 non-null    float64
 3   PetalLengthCm  150 non-null    float64
 4   PetalWidthCm   150 non-null    float64
 5   Species        150 non-null    int32  
dtypes: float64(4), int32(1), int64(1)
memory usage: 6.6 KB


In [10]:
df.describe()

Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm  \
count  150.000000     150.000000    150.000000     150.000000    150.000000   
mean    75.500000       5.843333      3.054000       3.758667      1.198667   
std     43.445368       0.828066      0.433594       1.764420      0.763161   
min      1.000000       4.300000      2.000000       1.000000      0.100000   
25%     38.250000       5.100000      2.800000       1.600000      0.300000   
50%     75.500000       5.800000      3.000000       4.350000      1.300000   
75%    112.750000       6.400000      3.300000       5.100000      1.800000   
max    150.000000       7.900000      4.400000       6.900000      2.500000   

          Species  
count  150.000000  
mean     1.000000  
std      0.819232  
min      0.000000  
25%      0.000000  
50%      1.000000  
75%      2.000000  
max      2.000000

In [11]:
df.corr()
#corr() is used to find the pairwise correlation of all columns in the dataframe. 
#Any na values are automatically excluded. For any non-numeric data type columns in the dataframe it is ignored.

Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  \
Id             1.000000       0.716676     -0.397729       0.882747   
SepalLengthCm  0.716676       1.000000     -0.109369       0.871754   
SepalWidthCm  -0.397729      -0.109369      1.000000      -0.420516   
PetalLengthCm  0.882747       0.871754     -0.420516       1.000000   
PetalWidthCm   0.899759       0.817954     -0.356544       0.962757   
Species        0.942830       0.782561     -0.419446       0.949043   

               PetalWidthCm   Species  
Id                 0.899759  0.942830  
SepalLengthCm      0.817954  0.782561  
SepalWidthCm      -0.356544 -0.419446  
PetalLengthCm      0.962757  0.949043  
PetalWidthCm       1.000000  0.956464  
Species            0.956464  1.000000

# Train Test Split

In [12]:
X=df.drop(["Species",'Id'],axis=1)
#X will drop species column and will take other columns
#species is our target variable
y = df["Species"]

In [13]:
from sklearn.preprocessing import MinMaxScaler # every datpoint will be between (0 - 1)
sc = MinMaxScaler()
X = sc.fit_transform(X)
X[:10]

array([[0.22222222, 0.625     , 0.06779661, 0.04166667],
       [0.16666667, 0.41666667, 0.06779661, 0.04166667],
       [0.11111111, 0.5       , 0.05084746, 0.04166667],
       [0.08333333, 0.45833333, 0.08474576, 0.04166667],
       [0.19444444, 0.66666667, 0.06779661, 0.04166667],
       [0.30555556, 0.79166667, 0.11864407, 0.125     ],
       [0.08333333, 0.58333333, 0.06779661, 0.08333333],
       [0.19444444, 0.58333333, 0.08474576, 0.04166667],
       [0.02777778, 0.375     , 0.06779661, 0.04166667],
       [0.16666667, 0.45833333, 0.08474576, 0.        ]])

In [14]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size = 0.25, random_state = 23)

In [15]:
X_train.shape

(112, 4)

In [16]:
X_test.shape

(38, 4)

# KNN Classification

In [17]:
from sklearn.metrics import accuracy_score
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import confusion_matrix
knn = KNeighborsClassifier()    #classifier= KNeighborsClassifier(n_neighbors=5, metric='minkowski', p=2 )  
knn.fit(X_train, y_train)   #classifier.fit(x_train, y_train)  
y_pred = knn.predict(X_test) ##Predicting the test set result  y_pred= classifier.predict(x_test)
accuracy = accuracy_score(y_test, y_pred)
print("KNN Default score is", accuracy)

KNN Default score is 0.9736842105263158


In [18]:
#from sklearn.metrics import confusion_matrix
#Creating the Confusion Matrix:
#Now we will create the Confusion Matrix for our K-NN model to see the accuracy of the classifier. Below is the code for it:
#Creating the Confusion matrix  
#from sklearn.metrics import confusion_matrix  
#cm= confusion_matrix(y_test, y_pred)  

In [19]:
# Confusion matrix
cm = confusion_matrix(y_test, y_pred)
print(cm)

[[16  0  0]
 [ 0 11  0]
 [ 0  1 10]]


# Logistic Regression Classification

In [20]:
from sklearn.linear_model import LogisticRegression
log_reg = LogisticRegression()
log_reg.fit(X_train, y_train)
y_pred = log_reg.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print("Logistic Regression Default score is", accuracy)

Logistic Regression Default score is 0.9473684210526315


In [21]:
cm = confusion_matrix(y_test, y_pred)
print(cm)

[[16  0  0]
 [ 0 10  1]
 [ 0  1 10]]


# Tune Model with GridSearchCV¶

In [22]:
log_reg = LogisticRegression()
log_reg_params = {"C":[0.001, 0.01, 0.1, 1, 5, 10, 100, 1000]}
gridcv_model = GridSearchCV(log_reg, log_reg_params, cv = 10).fit(X_train, y_train)
gridcv_model.best_params_

NameError: name 'GridSearchCV' is not defined

In [23]:
log_reg_tuned = LogisticRegression(C=1000)
log_reg_tuned.fit(X_train, y_train)
y_pred = log_reg_tuned.predict(X_test)
accuracy_tuned = accuracy_score(y_test, y_pred)
print("Logistic Regression Tuned Score :", accuracy_tuned)

Logistic Regression Tuned Score : 0.9736842105263158


In [24]:
cm = confusion_matrix(y_test, y_pred)
print(cm)

[[16  0  0]
 [ 0 11  0]
 [ 0  1 10]]


# SVM Classification

In [25]:
from sklearn.svm import SVC
svc = SVC()
svc.fit(X_train, y_train)
y_pred = svc.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print("SVM Default Score is", accuracy)

SVM Default Score is 0.9736842105263158


In [26]:
cm = confusion_matrix(y_test, y_pred)
print(cm)

[[16  0  0]
 [ 0 11  0]
 [ 0  1 10]]


# Tune Model with GridSearchCV

In [27]:
svc = SVC()
svc_params = {"kernel": ["rbf", "linear"],
             "C": [0.001, 0.01, 0.1, 1, 10, 50, 100],}
gridcv_model = GridSearchCV(svc, svc_params, cv = 10).fit(X_train, y_train)
gridcv_model.best_params_

NameError: name 'GridSearchCV' is not defined

In [28]:
svc_tuned = SVC(C=0.1, kernel = "rbf").fit(X_train, y_train)
y_pred = svc_tuned.predict(X_test)
accuracy_tuned = accuracy_score(y_test, y_pred)
print("SVC Tuned Score is", accuracy_tuned)

SVC Tuned Score is 0.9736842105263158


In [29]:
cm = confusion_matrix(y_test, y_pred)
print(cm)

[[16  0  0]
 [ 0 11  0]
 [ 0  1 10]]


# Naive Bayes Classification

In [30]:
from sklearn.naive_bayes import GaussianNB
nb = GaussianNB()
nb.fit(X_train, y_train)
y_pred = nb.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print("Naive Bayes Default Model Score is : ", accuracy)

Naive Bayes Default Model Score is :  0.9473684210526315


In [31]:
cm = confusion_matrix(y_test, y_pred)
print(cm)

[[16  0  0]
 [ 0 10  1]
 [ 0  1 10]]


# Random Forest Classification

In [32]:
from sklearn.ensemble import RandomForestClassifier
rf = RandomForestClassifier(random_state = 23)
rf.fit(X_train, y_train)
y_pred = rf.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print("Random Forest Default Model Score is : ", accuracy)

Random Forest Default Model Score is :  0.9736842105263158


In [33]:
cm = confusion_matrix(y_test, y_pred)
print(cm)

[[16  0  0]
 [ 0 11  0]
 [ 0  1 10]]


# Tune Model with GridSearchCV

In [34]:
rf = RandomForestClassifier(random_state = 23)
rf_params = {"n_estimators":[10,50,100,500,1000,2000],
            "criterion": ["gini", "entropy"],
            "min_samples_split": [2,5,10],
            "min_samples_leaf": [1,2,5,10,20,50]}
gridcv_model = GridSearchCV(rf, rf_params, cv = 10).fit(X_train, y_train)
gridcv_model.best_params_

NameError: name 'GridSearchCV' is not defined

In [35]:
rf_tuned = RandomForestClassifier(criterion = "gini", 
                                  min_samples_leaf = 2, min_samples_split = 5,
                                 n_estimators = 10, random_state = 23).fit(X_train, y_train)
y_pred = rf_tuned.predict(X_test)
accuracy_tuned = accuracy_score(y_test, y_pred)
print("Random Forest Tuned Model Score is : ", accuracy_tuned)

Random Forest Tuned Model Score is :  0.9736842105263158


In [36]:
cm = confusion_matrix(y_test, y_pred)
print(cm)

[[16  0  0]
 [ 0 11  0]
 [ 0  1 10]]


# Compare All Classification Models

Model Default Score Tuned Score
KNN ( Default Model Score : 0.97 - Tuned Model Score : 0.97 )

Logistic Regression ( Default Model Score : 0.95 - Tuned Model Score : 0.97 )

SVM ( Default Model Score : 0.97 - Tuned Model Score : 0.97 )

Naive Bayes ( Default Model Score : 0.94 - Tuned Model Score : )

Random Forest ( Default Model Score : 0.97 - Tuned Model Score : 0.97 )